march_hardware/src/Encoder.cpp

```diff
@@ -1,122 +1,133 @@
   return return_byte.i;
 }
 
-int32_t Encoder::RadtoIU(double rad)
-{
-  return static_cast<int32_t>(rad * totalPositions / (2 * M_PI) + zeroPositionIU);
-}
-
-double Encoder::IUtoRad(int32_t iu)
+int32_t Encoder::fromRad(double rad) const
 {
-  return (iu - zeroPositionIU) * 2 * M_PI / totalPositions;
+  return (rad * this->total_position_ / PI_2) + this->zero_position_iu_;
 }
 
-void Encoder::setSlaveIndex(int slaveIndex)
+double Encoder::toRad(int32_t iu) const
 {
-  this->slaveIndex = slaveIndex;
+  return (iu - this->zero_position_iu_) * PI_2 / this->total_position_;
 }
 
-int Encoder::getSlaveIndex() const
+bool Encoder::isWithinHardLimitsIU(int32_t iu) const
 {
-  return this->slaveIndex;
+  return (iu > this->lower_limit_iu_ && iu < this->upper_limit_iu_);
 }
 
-bool Encoder::isWithinHardLimitsIU(int32_t positionIU)
+bool Encoder::isWithinSoftLimitsIU(int32_t iu) const
 {
-  return (positionIU > this->lowerHardLimitIU && positionIU < this->upperHardLimitIU);
+  return (iu > this->lower_soft_limit_iu_ && iu < this->upper_soft_limit_iu_);
 }
 
-bool Encoder::isWithinSoftLimitsIU(int32_t positionIU)
+bool Encoder::isValidTargetIU(int32_t current_iu, int32_t target_iu) const
 {
-  return (positionIU > this->lowerSoftLimitIU && positionIU < this->upperSoftLimitIU);
-}
-
-bool Encoder::isValidTargetIU(int32_t currentIU, int32_t targetIU)
-{
-  if (this->isWithinSoftLimitsIU(targetIU))
+  if (target_iu <= this->lower_soft_limit_iu_)
   {
-    return true;
+    return target_iu >= current_iu;
   }
 
-  if (currentIU >= this->getUpperSoftLimitIU())
+  if (target_iu >= this->upper_soft_limit_iu_)
   {
-    return (targetIU <= currentIU) && (targetIU > this->getLowerSoftLimitIU());
+    return target_iu <= current_iu;
   }
 
-  if (currentIU <= this->getLowerSoftLimitIU())
-  {
-    return (targetIU >= currentIU) && (targetIU < this->getUpperSoftLimitIU());
-  }
+  return true;
+}
 
-  return false;
+void Encoder::setSlaveIndex(int slave_index)
+{
+  this->slave_index_ = slave_index;
 }
 
 int32_t Encoder::getUpperSoftLimitIU() const
 {
-  return upperSoftLimitIU;
+  return this->upper_soft_limit_iu_;
 }
 
 int32_t Encoder::getLowerSoftLimitIU() const
 {
-  return lowerSoftLimitIU;
+  return this->lower_soft_limit_iu_;
 }
 
 int32_t Encoder::getUpperHardLimitIU() const
 {
-  return upperHardLimitIU;
+  return this->upper_limit_iu_;
 }
 
 int32_t Encoder::getLowerHardLimitIU() const
 {
-  return lowerHardLimitIU;
+  return this->lower_limit_iu_;
+}
+
+size_t Encoder::calculateTotalPositions(size_t number_of_bits)
+{
+  if (number_of_bits < Encoder::MIN_RESOLUTION || number_of_bits > Encoder::MAX_RESOLUTION)
+  {
+    throw error::HardwareException(error::ErrorType::INVALID_ENCODER_RESOLUTION,
+                                   "Encoder resolution of %d is not within range [%ld, %ld]", number_of_bits,
+                                   Encoder::MIN_RESOLUTION, Encoder::MAX_RESOLUTION);
+  }
+  return std::pow(2, number_of_bits) - 1;
```

Comment:
Why do - 1? The encoder can also have a value of 0. Would not make a lot of difference but just for the sake of clarity.